### DISCUSSION TITLE
Not sure how to clean up old junk

### DISCUSSION THREAD

**Discussion by peeeeter** [arxanas/git-branchless#1218] [Q&A] [answered]
### Subcommand

hide

### Description

I forgot how this worked (with hg/branchless style) but I now have this very long and completely irrelevant history `git-branchless sl` on main:

What is the best practice for just getting to showing "main" only when you are on the master/main branch and not on a untracked or dev branch?

```
⋮ ✕ 323336d 28d (manually hidden) <REDACT>
⋮ ┃
⋮ ✕ 064bcdf 27d (manually hidden) <REDACT>
⋮
✕ 29f4270 7d (manually hidden) <REDACT>
┣━┓
┃ ✕ 7b712b2 7d (manually hidden) <REDACT>
┃ ┃
┃ ✕ 07a1e1f 7d (manually hidden) <REDACT>
┃
✕ 9c082af 7d (manually hidden) <REDACT>
┣━┓
⋮ ✕ 20673ac 7d (manually hidden) <REDACT>
⋮
✕ 15aaffa 7d (manually hidden) <REDACT>
┣━┓
⋮ ✕ 2d70658 7d (manually hidden) <REDACT>
⋮
✕ 5106361 7d (manually hidden) <REDACT>
┣━┓
┃ ✕ e0f8483 6d (manually hidden) <REDACT>
┃ ┃
┃ ✕ 1163042 6d (manually hidden) <REDACT>
┃ ┃
┃ ✕ 5e3ab12 6d (manually hidden) <REDACT>
┃
✕ e936612 5d (manually hidden) <REDACT>
┣━┓
┃ ✕ c985184 5d (manually hidden) <REDACT>
┃
✕ 57116b4 5d (manually hidden) <REDACT>
┣━┓
⋮ ✕ 5e05619 5d (manually hidden) <REDACT>
⋮
✕ 57f1c43 5d (manually hidden) <REDACT>
┣━┓
⋮ ✕ 44ad6cb 4d (manually hidden) <REDACT>
⋮ ┃
⋮ ✕ 257421d 28m (manually hidden) <REDACT>
⋮ ┃
⋮ ✕ dca73d4 28m (manually hidden) <REDACT>
⋮
◆ d61eff9 16m (ᐅ main) <REDACT>
```

**Comment by arxanas** [comment]
Generally speaking, hidden commits are not visible in the smartlog unless they have a visible descendant. In your case, this might mean that a branch is pointing into those commit stacks. You could delete the (local) branch in that case. See also https://github.com/arxanas/git-branchless/wiki/Command:-git-smartlog#which-commits-are-visible

You can provide a general [revset](https://github.com/arxanas/git-branchless/wiki/Reference:-Revsets) to `git-branchless smartlog` to print it instead and hide/show only some branches. Also see https://github.com/arxanas/git-branchless/wiki/Configuration#branchlesssmartlogdefaultrevset to configure the default revset for the smartlog. At work, I show all branches except for a certain long-lived one in this way (something like `draft() - ::branchname`, I don't remember).

**Comment by peeeeter** [comment]
Ah, thanks, but yes, I think that's my point.  In a repo, most of commits are ancestors of main, and if this is marked as a default in a branchless style smartlog, then you don't expect it to list all of them.  In this case, it's showing all my commits, since they are all ancestors of main.  I tried hiding them manually, without success.  Is this a bug?

**Reply by arxanas** [reply]
In your example, there are several commits that are not ancestors of main:

```
✕ 57f1c43 5d (manually hidden) <REDACT>
┣━┓
⋮ ✕ 44ad6cb 4d (manually hidden) <REDACT>
⋮ ┃
⋮ ✕ 257421d 28m (manually hidden) <REDACT>
⋮ ┃
⋮ ✕ dca73d4 28m (manually hidden) <REDACT>
⋮
◆ d61eff9 16m (ᐅ main) <REDACT>
```

in this case, `57f1c43` is an ancestor of `main`, but not `44ad6cb`. Because of this, `44ad6cb` is displayed in the smartlog. In such cases, the merge-base/greatest-common-ancestor with `main` is also displayed. But `44ad6cb` should not be visible unless there's a branch pointing to it or something like that.

Can you paste the result of `git branchless bug-report` here? It should include redacted smartlogs for examination.

**Reply by peeeeter** [reply]
[branchless_report.txt](https://github.com/arxanas/git-branchless/files/14320454/branchless_report.txt)


Thanks for taking a look!

**Reply by arxanas** [reply]
The smartlog shows that there are branches pointing to those commits (or their descendants), so they are visible by default:

```
: x 064bcdf 37d (manually hidden) (redacted-ref-1) xxxx xxxxxxxxx xxxx xxx xxxxxx xxxxxxxxxxxxx xx xxxxx xxxxxxx
```

**Comment by peeeeter** [comment]
Ah, I see.  I thought if I manually hid those refs, they would go away,
regardless of git branches tagging them. Thanks for clarifying.

**Comment by peeeeter** [comment]
Oh and there is a "-D" option for hide.  Okay, that's what i needed to do.

**Reply by arxanas** [reply]
`-D` is the default for `git hide` in the latest version of git-branchless.

**Comment by peeeeter** [comment]
It would be cool to have a configuration to help do this for you when it
gets in this state for those of us now living in a "create branch for
issue" github world.  It would basically just cleanup and hide branches for
committed PRs.

**Reply by arxanas** [reply, marked as answer]
I usually use [`git sync`](https://github.com/arxanas/git-branchless/wiki/Command:-git-sync) to clean up my merged branches (but it doesn't currently work for squash-merges).

**Reply by peeeeter** [reply]
git sync is the magic that makes this work the way i want.  Thanks so much.  Hope to help/contribute sometime.... This is the best.

**Reply by peeeeter** [reply]
I might even go so far as to suggest adding it as a hint for "hide" in the docs.  feel free to use my example of  "if you end up with this mess.... "
